# Geocoding Organisations

For a number of planned analyses, we require the latitude and longitude of the organisations in the data, not just their names and addresses. To do this we will first use a geocoding service to identify lat lng from address strings.

In [466]:
import pickle
import ratelim
import json
import numpy as np
import os
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

from geopy.geocoders import GoogleV3
from ipywidgets import FloatProgress
from IPython.display import display

key = os.environ.get('GOOGLE_MAPS_KEY')

In [3]:
# Google geocoder
geocoder = GoogleV3(api_key=key)

In [4]:
# Read in config file with DB params
with open('../scripts/config.json') as f:
    conf = json.load(f)
    
# Define a connection string
conn_string = 'host={} dbname={} user={} password={}'.format(conf.get('host'),
                                                             conf.get('database'),
                                                             conf.get('user'),
                                                             conf.get('passw'))

# Create a connection object
conn = psycopg2.connect(conn_string)

# SQL string that unpacks nested JSON arrays
sql_str = """
SELECT id, name, addresses from gtr.organisations;
"""

# chunksize returns an iterator that reads chunk number of rows at a time
df = pd.read_sql(sql_str,
                 conn)

In [5]:
def country(x):
    try:
        return x.get('address', None)[0].get('country', None).lower()
    except:
        return None

In [356]:
# API rate limits with billing enabled
@ratelim.greedy(5, 1)
def geocode(address_dict):
    # Get the list out of the dict
    address_list = address_dict['address']
    
    # If no country need to assume UK
    if address_list and address_list[0].get('country', 'uk').lower() in uk_names:
        try:
            pc = address_dict['address'][0]['postCode']
            return geocoder.geocode(pc)
        except:
            address = address_str(address_dict)
            if address:
                return geocoder.geocode(address)
            else:
                return "here"
    else:
        address = address_str(address_dict, uk=False)
        if address:
            return geocoder.geocode(address)
        else:
            return "here2"

In [7]:
def postcode(address_entry):
    try:
        return address_entry['address'][0].get('postCode', None)
    except:
        return None

In [95]:
def address_str(address_dict, uk=True):
    
    uk_address_keys = [
    'line1',
    'line2',
    'line3',
    'county',
    'postCode',
    ]
    
    other_address_keys = [
        'line1',
        'county',
        'country',
    ]
    
    if uk:
        address_keys = uk_address_keys
    else:
        address_keys = other_address_keys
    
    l = []
    address_list = address_dict.get('address', None)
    if address_list is None:
        return None
    try:
        if address_list[0] is not None:
            [l.append(address_list[0][key]) if key in address_keys else '' for key in address_list[0]]            
    except:
        return None
    
    return ", ".join(l)

In [9]:
def check_address(address_dict):
    address_list = address_dict.get('address', None)
    if address_list is None:
        return False
    try:
        if address_list[0] is not None:
            return True
    except:
        return False

In [390]:
# Function to get tuple by position, returning
# None where there isn't a valid entry
def get_tuple_value(x, pos):
    try:
        return x[pos]
    except TypeError:
        return None

In [10]:
# All the ways people put down UK in the country field
uk_names = ['united kingdom',
 'uk',
 'wales',
 'england',
 'scotland',
 'united kingom',
 'northern ireland',
 'united kindgom',
 'buckinghamshire',
 'yorkshire',
 'london',
 'u.k.',
 'united ,kingdom',
 'berkshire',
 'jersey',
 ' united kingdom',
 'u.k']

In [11]:
df['postcode'] = df.addresses.map(postcode)
n_postcodes = df.postcode.count()
n_no_postcodes = len(df.postcode) - n_postcodes
print('{} organisations with postcodes, {} without'.format(n_postcodes, n_no_postcodes))

18266 organisations with postcodes, 7299 without


We will be unable to geocode 7,299 organisations by using only a postcode in the geocode request. Lets check how many of those without postcodes, also lack other address details. For instance, organisations from outside the UK would lack a postcode, but can still be geocoded using a string lookup.

In [12]:
# Full dataframe check
df.addresses.map(check_address).value_counts()

True     21205
False     4360
Name: addresses, dtype: int64

In [13]:
# Subset of those without postcodes
df[df.postcode.isnull()].addresses.map(check_address).value_counts()

False    4360
True     2939
Name: addresses, dtype: int64

Of the 7,299 organisations without a postcode, 4,360 also have no address data. This leaves 2,939 organisations with no postcode but some data in the address field. Therefore the maximum number of geocoded organisations will be 21,205. When geocoding, we will default to the use of a postcode where possible, and fallback to an address string where a postcode isn't present.

In [387]:
"""latlng = []
num_addresses = df.addresses.shape[0]
f = FloatProgress(min=0, max=num_addresses)
display(f)

for address in df.addresses:
    g = geocode(address)
    try:
        latlng.append((g.latitude, g.longitude))
    except AttributeError:
        latlng.append(None)
    f.value += 1
"""   

In [394]:
# Save as a pickled object
pickle.dump( latlng, open( "latlng.p", "wb" ) )

In [474]:
df_latlng = pd.DataFrame(np.asarray(latlng))

In [480]:
df_result = df.merge(df_latlng, left_index=True, right_index=True)
df_result.rename(columns={0: 'latlng'}, inplace=True)
df_result = df_result[df_result.latlng.notnull()]
df_result[['lat', 'lng']] = df_result.latlng.apply(pd.Series)
df_result.drop(['postcode', 'addresses', 'latlng'], axis=1, inplace=True)

In [481]:
df_result

id  \
0      BAB46AB6-90FC-4145-B885-0153BEA15A73   
2      BCD9069A-6739-4039-86DB-042CB13254CB   
3      B69FACF5-5AB9-4D97-88CD-0A38C32935FF   
4      B76FE143-1E07-40B0-8932-028207296A78   
5      B80095E4-9D0F-4398-8A0E-07EB418EFFF6   
6      B966BD7A-F468-4C74-A694-0842CB8F884F   
8      BA32CA96-1A26-415F-B3B2-0DC1DDFCE74F   
9      BA59F886-138D-41EB-B9B1-0C79E45C61D3   
10     BA7D8BB2-4B17-4D57-AC78-001BA0F40F77   
11     C059DB15-1836-430B-A993-0759206ABBB8   
13     C066B655-E1BA-447A-A432-01FB6725E320   
14     C2B3EE68-0D08-4321-964A-02108F67A3F8   
15     C30DF86F-9DD8-4A78-97BA-067527E5E391   
18     C42E852D-5D31-4DE8-8E91-04CF66DA9DE2   
19     C59C7E94-5DD5-4DCA-8586-09189A61F279   
20     C8A76B9E-4F39-490B-BCCF-009F592A756C   
21     C9069C42-693D-4C9C-B66B-0D968C7277AA   
22     C9424089-9FC7-4BD3-8C9A-0E38B885E2CE   
23     D04848C3-4132-45F1-8841-06DB92EB1E87   
24     D068E964-7B96-42CA-9F67-00BE1677AE7F   
25     D080060B-4CE4-495E-8C47-07BFD74CB8E0   
26     D3BC1D1D-CEC8-4A05-A0B0-0D6167F5A0B0   
27     D3E9EF03-CF8E-4251-BCF4-01E09BFADDA8   
28     D42BD99F-AD31-43D6-8FBE-0B96F1E266D3   
29     CA10F5D0-8E26-4790-BF21-025009DDA322   
32     D5542D2A-CB08-48B1-91A6-03A59EA2B30F   
33     D5F4323F-A8DF-4869-B0F0-0EB45A7A7F95   
34     D6C3234E-5048-4AA0-A116-0CAC51569B2D   
36     D72C440C-6E4E-4FEE-94D9-0652BB970C3B   
38     DFFCC0DB-5473-43BE-AC92-06CA6F272331   
...                                     ...   
25530  621FCA67-9259-4CBE-AECE-D7D4C4133C54   
25531  82DA3271-1C71-474F-BF1B-D122CF76DAFF   
25532  839FAFB1-DF5C-46B3-96AD-CF18D4BC7AFC   
25533  84E231EE-EC82-44AC-9838-C30021DDA51A   
25534  84F84A6B-60BD-4285-B4D0-E334D8B8938A   
25535  99A606D5-9E6E-4968-B976-C1A25D67F9AF   
25536  99F15E99-5A6E-4F9E-911B-C5F8E010BAD8   
25537  9B99C7C4-8160-46DC-89F7-C2F36B3767E2   
25538  9D1B7194-491E-42B4-A4AF-C1D4DBAA3CCE   
25539  B249150F-44BF-48CE-8135-D42FF59D0344   
25541  A397012C-D94E-45CA-A7E6-BD6F074B7028   
25544  AAAE0503-DE4B-47B6-8C4B-B46D0062D536   
25545  AAE3A59D-9701-4A7E-8EC4-BA03027E466E   
25547  ACBAEB96-C59A-433B-BC77-BAD807671AA4   
25548  AD07723C-4B8A-40DB-B09C-C83E40869CFD   
25549  AD1B5C1D-39E1-4A1F-A58F-D5F2DE64FE7F   
25550  AD4118A8-85A8-403C-8108-C3C6146D14C4   
25551  AD44F225-3942-418A-B036-C699233AD12A   
25552  ADCEEE0C-FB0E-4EED-89BB-C0FB6A253F0E   
25553  6E8646B8-5C71-4DE5-9164-FC5E68CAE56C   
25554  7345A886-BEC0-4B87-9EE2-E5B93F7938AF   
25555  6716AB88-0CC4-464D-9A37-FAEFDBF28FB7   
25557  6917202A-7D08-4660-964C-EF12D6E0F8E3   
25558  82F16418-FF71-4AB8-9185-FDBCE2CFB46E   
25559  861D241C-0324-4689-87EF-F91D681ECD16   
25560  86952E0A-D64A-4CE7-A0ED-EE2C1F02A07F   
25561  88C1DE3B-3852-4F78-8685-F80B4D748476   
25562  7AA4FAF1-DD16-466B-8852-EE37FB8A85F0   
25563  7CEF38BD-DE6D-45C1-9335-EC1B884E4146   
25564  7EC3733E-A3C2-4449-8889-F893E9C790E8   

                                              name        lat         lng  
0                      A-Tech Fabrications Limited  54.604095   -1.574308  
2                             Zurich Insurance plc  50.881574   -1.242902  
3                                 Limagrain UK Ltd  53.482100   -0.248484  
4                  China Earthquake Administration  39.977022  116.306198  
5                         Newmills Engineering Ltd  51.504765   -2.484122  
6                            Maastricht University  50.844288    5.688604  
8                                    Synthomer Ltd  51.784388    0.119136  
9                      Committee on Climate Change  51.494349   -0.140087  
10                          Agility Global Limited  51.650564   -0.149673  
11                                    Openreach BT  51.515741   -0.097966  
13                             Emote Games Limited  50.876386   -1.240271  
14                        Total Scientific Limited  52.094562    0.155632  
15              Omega Engineering Services Limited  51.901120   -2.097201  
18                               Swanshurst School  52.429668 

In [482]:
engine = create_engine('postgresql://{}:{}@{}/{}'.format(conf.get('user'),
                                                         conf.get('passw'),
                                                         conf.get('host'),
                                                         conf.get('database')))

conn = engine.connect()

In [483]:
df_result.to_sql('orgs_latlng',
                 conn,
                 schema='gtr',
                 if_exists='replace',
                 index=False)